src/tracking.jl

```diff
@@ -63,35 +63,45 @@ is_tracking_lmo(lmo::TrackingLMO) = true
 
 TrackingLMO(lmo) = TrackingLMO(lmo, 0)
 
-"""
-    tracking_trajectory_callback(storage)
 
-Similar to `trajectory_callback`, appropriate for tracking vanilla frank_wolfe without lmo caching.
-Pushes the state at each iteration to the passed storage.
-The state data is only the 5 first fields, gamma and 3 call counters, usually
+"""
+A function acting like the passed callback,
+    but adding the state to the storage variable.
+    The state data is only the 5 first fields, gamma and 3 call counters, usually
 `(t, primal, dual, dual_gap, time, gamma, function_calls, gradient_calls, lmo_calls)`
 """
-function tracking_trajectory_callback(storage)
-    return function tracking_push_trajectory!(state)
-        base_tuple = Tuple(state)[1:5]
-        complete_tuple = tuple(base_tuple..., state.gamma, state.f.counter, state.grad!.counter, state.lmo.counter)
-        return push!(storage, complete_tuple)
-    end
+mutable struct TrackingCallback{C} <: Function
+    callback::C
+    storage::Vector
+end
+
+TrackingCallback(callback) = TrackingCallback(callback, [])
+TrackingCallback() = TrackingCallback(state->false, [])
+
+function (tc::TrackingCallback)(state)
+    base_tuple = Tuple(state)[1:5]
+    complete_tuple = tuple(base_tuple..., state.gamma, state.f.counter, state.grad.counter, state.lmo.counter)
+    push!(tc.storage, complete_tuple)
+    return tc.callback(state)
 end
 
 """
-    tracking_cached_trajectory_callback(storage)
-    
-Similar to `trajectory_callback`, appropriate for tracking progress with cached LMOs, e.g. lazified_conditional_gradient.
-Pushes the state at each iteration to the passed storage.
-The state data is only the 5 first fields, gamma and 3 call counters, usually
+A function acting like the passed callback for cached LMOs,
+    but adding the state to the storage variable.
+    The state data is only the 5 first fields, gamma and 3 call counters, usually
 `(t, primal, dual, dual_gap, time, gamma, function_calls, gradient_calls, lmo_calls)`
-For this callback, the counter method is accessed through the inner LMO inside its CacheLMO wrapper 
 """
-function tracking_cached_trajectory_callback(storage)
-    return function tracking_push_trajectory!(state)
-        base_tuple = Tuple(state)[1:5]
-        complete_tuple = tuple(base_tuple..., state.gamma, state.f.counter, state.grad!.counter, state.lmo.inner.counter)
-        return push!(storage, complete_tuple)
-    end
+mutable struct TrackingCachedCallback{C} <: Function
```

Comment:
I don't think it is necessary to turn this closure into a struct.
This can be fine but requires giving a concrete type to `storage`, so having a concrete eltype on the `::Vector`

Reply:
I can see the interest but would stay with the closure

Reply:
I used this mutable struct in a similar way than the TrackingGradient, Function, LMO versions before, but in the new commits I have adapted the closure to also take a possible boolean callback as an additional argument to the storage, which preserves the old callback structure with external 'storage' variable